hail/python/hailtop/aiogoogle/client/compute_client.py

```diff
@@ -77,13 +82,50 @@ async def list(self, path: str, *, params: MutableMapping[str, Any] = None, **kw
         return PagedIterator(self, path, params, kwargs)
 
     async def create_disk(self, path: str, *, params: MutableMapping[str, Any] = None, **kwargs):
-        return await request_with_wait_for_done(self.post, path, params, **kwargs)
+        return await self._request_with_zonal_operations_response(self.post, path, params, **kwargs)
 
     async def attach_disk(self, path: str, *, params: MutableMapping[str, Any] = None, **kwargs):
-        return await request_with_wait_for_done(self.post, path, params, **kwargs)
+        return await self._request_with_zonal_operations_response(self.post, path, params, **kwargs)
 
     async def detach_disk(self, path: str, *, params: MutableMapping[str, Any] = None, **kwargs):
-        return await request_with_wait_for_done(self.post, path, params, **kwargs)
+        return await self._request_with_zonal_operations_response(self.post, path, params, **kwargs)
 
     async def delete_disk(self, path: str, *, params: MutableMapping[str, Any] = None, **kwargs):
-        return await request_with_wait_for_done(self.delete, path, params, **kwargs)
+        return await self.delete(path, params=params, **kwargs)
+
+    async def _request_with_zonal_operations_response(self, request_f, path, params: MutableMapping[str, Any] = None, **kwargs):
+        assert 'params' not in kwargs
+
+        async def request_and_wait():
+            local_params = copy.deepcopy(params)
+
+            if local_params is None:
+                local_params = {}
+
+            request_uuid = str(uuid.uuid4())
+            if 'requestId' not in local_params:
+                local_params['requestId'] = request_uuid
+
+            resp = await request_f(path, params=local_params, **kwargs)
+            operation_id = resp['id']
+            zone = resp['zone'].rsplit('/', 1)[1]
+
+            while True:
+                result = await self.post(f'/zones/{zone}/operations/{operation_id}/wait')
+                if result['status'] == 'DONE':
+                    http_error_status_code = result.get('httpErrorStatusCode')
+                    http_error_message = result.get('httpErrorMessage')
+                    errors = None
+
+                    error = result.get('error')
+                    if error:
+                        errors = error['errors']
+                        errors = [GCPError(e['code'], e['message']) for e in errors]
+
+                    if http_error_status_code is not None:
+                        raise GCPOperationError(http_error_status_code, http_error_message, errors, result)
```

Comment:
Hmm. I don't like the correlation between `errors is not None` and `http_error_status_code is not None`. I don't think we actually care about the HTTP status code and error message. If there are errors, then something went wrong, right?

What do you think of this?
```python3
error = result.get('error')
if error:
    raise GCPOperationError([GCPError(e['code'], e['message']) for e in error['errors']], result)
```

Reply:
Or maybe:
```python3
error = result.get('error')
if error:
    assert result.get('httpErrorStatusCode') is not None
    assert result.get('httpErrorMessage') is not None
    raise GCPOperationError(result['httpErrorStatusCode'],
                            result['httpErrorMessage'],
                            [GCPError(e['code'], e['message']) for e in error['errors']],
                            result)
```